functions/go/delete-annotations/main.go

```diff
@@ -0,0 +1,133 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"sigs.k8s.io/kustomize/kyaml/fn/framework"
```

Comment:
@yuwenma Should we ask users to try to use the SDK in the kpt-functions-sdk repo?
The logic in this function is relatively simple, I'd suggest to implement the function with the new SDK. WDYT?

Reply:
Totally agree we should push users using the new SDK. Also, deleting annotations should be a standalone function (at least not in the curated catalog) because this is an imperative operation and theoretically should only be called by `kpt fn eval`. If it is only called by `kpt fn eval` as one-time change, `git restore` could be an easier and cleaner solution.  @g-awmalik do you mind explaining more about the use cases? If this deleting annotation is called by `kpt fn render`, please make the change on existing `set-annotation` function. But that function uses the old kustomize/kyaml sdk and the majority of the code is encapsulated as a kustomize transformer so the change may not be easy to achieve. 
@g-awmalik anyway, let us know if `git` can fulfill the needs. If not, you can either make the change in set-annotation or file an issue and we will do the implementation for you.

Reply:
Calling this fn imperatively is the main use case and I don't think `git restore` achieves it since that may wipe out additional (necessary) changes as well. For me specifically, I needed fn to remove annotations before publishing resources for consumers. E.g. we add annotations to Gatekeeper Constraint resources to process them but those annotations are not useful for the customers using them and would like to take them out before publishing them on GitHub. Like a mentioned above, declarative execution is not required and imperative will work just fine.

Thanks for offering the help but I can't wait until May for this. What are options for getting this out? Can we push this to the registry without making it part of the catalog? Or is modifying `set-annotations` the only feasible option here?

Reply:
Also, I might add that deleting annotations using `set-annotations` may go against the guidelines for fn naming i.e. <verb>-<noun> and it may not be obvious to users at first that the fn supports deleting annotations as well.

Reply:
@yuwenma - tagging you as a reminder.

Reply:
Thank you for the context. I took a deeper look at the set-annotation function and the change would be non-trivial. Basically we need to refactor the code completely to use the new SDK and fork/extend the kustomize/api and kustomize/kyaml library to support object reset (create, update are supported but not delete). But I'm afraid we cannot plan this work before May.
As a short term quick solution, you can run the delete-annotation as executable in your local. Have you tried the --exec flag in `kpt fn eval`? The PR description said "needed this for myself ", you don't need to have the function be stored in the `gcr.io/kpt-fn`container registry. If you want to share the image with others, you can build it (Here's the [Dockerfile](https://github.com/GoogleContainerTools/kpt-functions-catalog/blob/master/build/docker/go/Dockerfile)) and upload the image to your own container registry. Would that help?

Reply:
Thanks @yuwenma. Is it OK if I publish this under [`contrib`](https://github.com/GoogleContainerTools/kpt-functions-catalog/tree/master/contrib)

Reply:
That's a good idea. If so, can you update this PR to use the new sdk as Mengqi suggested? Here's an example PR https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/769

Reply:
@yuwenma Can you explain why you think this function should not be added to the main repo? I think it's a critical and common requirement to remove annotations, so I don't really see a need to segregate it in `contrib/`.

Also not sure how this is "imperative-only." It works well in either imperative or declarative modes.

Reply:
Annotation is a horizontal function and we will extend set-annotation to add, update and delete annotations. So we don't plan to accept another function other than set-annotation in the curated catalog. 

"how this is imperative-only": Awais explains the use case as deleting the annotation as a one-off operation before publishing the kpt pkg. So the delete-annotation itself is not included in kptfile.

Reply:
> Annotation is a horizontal function and we will extend set-annotation to add, update and delete annotations. So we don't plan to accept another function other than set-annotation in the curated catalog.

Do you have an idea for the UX to do this? I do agree that we should probably just update the existing function.

> "how this is imperative-only": Awais explains the use case as deleting the annotation as a one-off operation before publishing the kpt pkg.

That is one possible use. It is not definitive. We should not use whether someone is invoking a function imperatively or declaratively as a criterion for inclusion in the catalog.

Reply:
```
Do you have an idea for the UX to do this? I do agree that we should probably just update the existing function.
```
I see two possibilities:
- we make the update to set-annotations as it is now and refactor the whole thing later (in May-ish timeframe)
- we publish delete-annotations to contrib for now and deprecate it when set-annotations is fully refactored

Reply:
> we make the update to set-annotations as it is now and refactor the whole thing later (in May-ish timeframe)

This is my preference, since it isn't really blocking to have the delete-annotations function published.

Reply:
@yuwenma - please share your thoughts.

Reply:
Refactoring the whole thing later is not a problem for me, but changing the set-annotation is not a easy task. I would say much harder than writing a delete function. 

- It needs to support not merely the metadata.annotations, but deleting other "annotation" fields like Deployment `spec.template.annotations` etc.  
- fork and change the third-party library kustomize/api fieldspec filter which has been highly encapsulated.
- it requires a new UX design even before refactoring and CaD team should own the UX design. 

Thus, I'd suggest publishing the delete-annotation to contrib which is much faster and easier. @g-awmalik

Reply:
As for the UX, I've assigned myself on #2524. We should design the UX for annotation, labels and probably other meta fields altogether and build a new user pattern in the sdk. I've seen too many risk conditions and would like to think more careful about it before rushing to the design.